### PR TITLE
Resolve money parsing exception in Retirement Defection Tracker #1316

### PR DESCRIPTION
**Pull request by sixlettervariables**
~~Adds Money::parse to deal with parsing monies in Retirement Defection Tracker (share calculations).~~ This bit should be done a different way ~~, but this resolves this for now.~~

This deals with net worth in retirement defection tracking by hoisting out the financial report calcs into their own class. No more parsing.

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/personnel/RetirementDefectionTracker.java [hunk 1]
```diff
@@ -99,15 +98,15 @@ public static Money getShareValue(Campaign campaign) {
         String financialReport = campaign.getFinancialReport();
         Money netWorth = Money.zero();
         try {
-            Pattern p = Pattern.compile("Net Worth\\D*(.*)");
+            Pattern p = Pattern.compile("Net Worth\\D*([\\d,.]+)");
```
Comment (neoancient) [inline MekHQ/src/mekhq/campaign/personnel/RetirementDefectionTracker.java]: An unescaped . matches anything, so [\\d,.]+` is the same as `.+`.

File: MekHQ/src/mekhq/campaign/finances/Money.java [hunk 2]
```diff
@@ -163,6 +162,10 @@ public String toAmountAndNameString() {
         return CurrencyManager.getInstance().getUiAmountAndNamePrinter().print(this.wrapped.toMoney(RoundingMode.HALF_EVEN));
     }
 
+    public static Money parse(String moneyStr) {
+        return new Money(BigMoney.parse(moneyStr));
```
Comment (neoancient) [inline MekHQ/src/mekhq/campaign/finances/Money.java]: The documentation for BigMoney#parse says it expects a String in the format of a three-letter currency code followed by a numeric value separated by one or more spaces. What happens when you don't have the code?
https://www.joda.org/joda-money/apidocs/org.joda.money/org/joda/money/BigMoney.html#parse(java.lang.String)

File: MekHQ/src/mekhq/campaign/personnel/RetirementDefectionTracker.java (hunk 1 above)
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/personnel/RetirementDefectionTracker.java]: You can thank night shift programming for this.

File: MekHQ/src/mekhq/campaign/finances/Money.java (hunk 2 above)
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/finances/Money.java]: It worked in testing, but I've decided to go another way with this that doesn't involve parsing strings.

File: MekHQ/src/mekhq/campaign/personnel/RetirementDefectionTracker.java [hunk 3]
```diff
@@ -99,15 +98,15 @@ public static Money getShareValue(Campaign campaign) {
...
             m.find();
-            netWorth = Money.of(Double.parseDouble(m.group(1)));
+            netWorth = Money.parse(m.group(1));
             if (campaign.getCampaignOptions().getSharesExcludeLargeCraft()) {
-                p = Pattern.compile("Large Craft\\D*(.*)");
+                p = Pattern.compile("Large Craft\\D*([\\d,.]*)");
                 m = p.matcher(financialReport);
                 if (m.find() && null != m.group(1)) {
-                    netWorth = netWorth.minus(Money.of(Double.parseDouble(m.group(1))));
+                    netWorth = netWorth.minus(Money.parse(m.group(1)));
```
[the remark this answers is not in the document]
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/personnel/RetirementDefectionTracker.java]: I resolved this in another way.

**Pull request merged**